ChronoStore/src/storywriter.cpp

```diff
@@ -0,0 +1,497 @@
+        }
+        attributeDataBuffer[attributeCount].startTimeStamp = eventBuffer[eventIndex].timeStamp;
+        attributeDataBuffer[attributeCount].endTimeStamp = eventBuffer[eventIndex + chunksize - 1].timeStamp;
+        attributeDataBuffer[attributeCount].datasetIndex = eventIndex;
+        attributeDataBuffer[attributeCount].totalChunkEvents = chunksize;
+        attributeCount++;
+    }
+
+    /*
+    * Create datatype for the ChunkMetadata attribute.
+    */
+    herr_t startError, endError,datasetIndexError , totalChunkEventsError, attrMaxError, attrStoryEventsError, attrMinError, attrError;
+    chunkmetaAttribute = H5Tcreate(H5T_COMPOUND, sizeof(ChunkAttr));
+    startError = H5Tinsert(chunkmetaAttribute, "startTimeStamp", HOFFSET(ChunkAttr, startTimeStamp), H5T_NATIVE_ULONG);
+    endError = H5Tinsert(chunkmetaAttribute, "endTimeStamp", HOFFSET(ChunkAttr, endTimeStamp), H5T_NATIVE_ULONG);
+    datasetIndexError = H5Tinsert(chunkmetaAttribute, "datasetIndex", HOFFSET(ChunkAttr, datasetIndex), H5T_NATIVE_UINT);
+    totalChunkEventsError = H5Tinsert(chunkmetaAttribute, "totalChunkEvents", HOFFSET(ChunkAttr, totalChunkEvents), H5T_NATIVE_UINT);
+    if(chunkmetaAttribute < 0 || endError < 0 || startError < 0 || totalChunkEventsError < 0 || datasetIndexError < 0){
+        if(attributeDataBuffer != NULL) free(attributeDataBuffer);
+        if(eventBuffer != NULL) free(eventBuffer);
+        if(chunkmetaAttribute >=0 ) H5Tclose(chunkmetaAttribute);
+        if(eventType >=0 )  H5Tclose(eventType);
+        if(storyDataset >= 0)   H5Dclose(storyDataset);
+        if(ChronicleH5 >= 0)  H5Fclose(ChronicleH5);
+        return -1;
+    }
+
+    if(H5Aexists_by_name(storyDataset, ".", ATTRIBUTE_CHUNKMETADATA, H5P_DEFAULT) > 0){
+        /*
+        * Open the attribute, get attribute id, type and space
+        */
+        attributeId = H5Aopen(storyDataset, ATTRIBUTE_CHUNKMETADATA, H5P_DEFAULT);
+        attributeType = H5Aget_type(attributeId);
+        attributeDataSpace = H5Aget_space(attributeId);
+        hsize_t attributeChunkMetadataDims = H5Sget_simple_extent_npoints(attributeDataSpace);
+        H5Sclose(attributeDataSpace);
+
+        hsize_t attributeDims[] = {attributeChunkMetadataDims + attributeCount};
+        extendChunkMetadataBuffer = (ChunkAttr *) malloc(attributeDims[0] * sizeof(ChunkAttr));
+
+        /*
+        * Read the attribute data into the buffer and delete the attribute
+        */
+        status = H5Aread(attributeId, attributeType, extendChunkMetadataBuffer);
+        if(status < 0 || attributeId < 0 || attributeType < 0 || extendChunkMetadataBuffer == NULL || attributeChunkMetadataDims == HSIZE_UNDEF){
+            if(attributeId >= 0)  H5Aclose(attributeId);
+            if(attributeType >= 0)  H5Tclose(attributeType);
+            if(extendChunkMetadataBuffer != NULL)   free(extendChunkMetadataBuffer);
+            if(attributeDataBuffer != NULL) free(attributeDataBuffer);
```

Comment:
Code readability would have been better if you could combine all error checks into a couple of functions and call those functions here. But, that is just a comment.

Reply:
added goto. Done